### PR TITLE
add support for initializer list

### PR DESCRIPTION
**Pull request by lemire**
Basically, we want to be able to do the following:

```C++
Roaring b = Roaring::bitmapOfList({1,2,4});
```

Instead of the awkward

```C++
Roaring  = Roaring::bitmapOf(3, 1, 2, 4);
```

### CODE REVIEW

**Comment by lemire** [comment]
@kosak Would you review ?

File: tests/cpp_random_unit.cpp [hunk 1]
```diff
@@ -482,7 +482,7 @@ DEFINE_TEST(random_doublecheck_test_64) {
 int main() {
     uint64_t seed = time(nullptr);
     srand(seed);
-    printf("Seed: %lu\n", seed);
+    printf("Seed: %llu\n", seed);
```
Comment (madscientist) [inline tests/cpp_random_unit.cpp:485]: It's unfortunate but this is not portable.  On some systems, `int` is 32bits and `long` and `long long` are 64bits.  On those systems, `uint64_t` often is an alias for `long`, not `long long`, and `%llu` fails if you have format string checking enabled.  See: https://godbolt.org/z/6zr1ndq86

 The new C++ `std::fmt` facility fixes all this and is much nicer than printf, but unfortunately it's not available in older versions of C++.  You will probably have to add a cast to `unsigned long long` if you want to avoid the format string warnings on all systems.

**Comment by kosak** [comment]
Yes, I'll review (I was about to propose something similar!)

File: cpp/roaring.hh [hunk 2]
```diff
@@ -112,6 +121,18 @@ public:
         return ans;
     }
 
+    /**
+     * Construct a bitmap from a list of uint32_t values.
+     * E.g., bitmapOf({1,2,3}).
+     */
+    static Roaring bitmapOf(std::initializer_list<uint32_t> l) {
```
Comment (kosak) [inline cpp/roaring.hh:128]: I have a slight concern with overloading `bitmapOf` (i.e. using the same name) in the context of people that like to use uniform initialization syntax (curly braces rather than parens).

Before this change, `bitmapOf({0})` would have been a valid call (to the `va_list` version), yielding a bitmap of size zero.
After this change, `bitmapOf({0})` now makes a bitmap of size 1, with its element 0 set.

I can't think of any other real-world examples that are likely to trip people up, but perhaps for the avoidance of confusion, the initializer_list version could be called something different like `bitmapOfList`.

File: tests/cpp_random_unit.cpp (hunk 1 above)
Comment (lemire) [reply tests/cpp_random_unit.cpp]: Done. See my last commit.

File: tests/cpp_random_unit.cpp [hunk 3]
```diff
@@ -378,7 +379,7 @@ DEFINE_TEST(random_doublecheck_test_64) {
         const Roaring64Map &right = roars[rand() % NUM_ROARS];
 
 #ifdef ROARING_CPP_RANDOM_PRINT_STATUS
-        printf("[%lu]: %llu %llu %llu\n", step,
+        printf("[%lu]: %" PRIu64 " %" PRIu64 " %" PRIu64 "\n", step,
                static_cast<unsigned long long>(left.cardinality()),
                static_cast<unsigned long long>(right.cardinality()),
                static_cast<unsigned long long>(out.cardinality()));
```
Comment (kosak) [inline tests/cpp_random_unit.cpp:385]: Now that you're using `PRIu64` there should NOT be casts here any more.

File: cpp/roaring.hh [hunk 4]
```diff
@@ -61,6 +61,15 @@ public:
...
 
+    /**
+     * Construct a bitmap from an initializer list.
+     */
+    Roaring(std::initializer_list<uint32_t> l) : Roaring() {
+        for(uint32_t x : l) {
+            add(x);
+        }
+    }
+
```
Comment (kosak) [inline cpp/roaring.hh:72]: I had one more (long) comment here and I can't find now so you might see it twice.

Basically I am concerned about the interaction between initializer lists and uniform initialization syntax.

Consider a hypothetical future world where you have (say) a two-arg constructor and an initializer list and a customer who likes uniform initialization syntax.

Library:
```
Roaring(int a, int b) {...}
Roaring(std::initializer_list<uint32> l) {...}
```

client:
```
// Note curly braces. This is valid syntax for the two-arg constructor but the compiler will
// prefer the initializer_list constructor.
Roaring r{5, 100};  // calls initializer_list constructor
```

I see two valid approaches:
1. Don't worry too much about the vagaries of uniform initialization syntax
2. Take away the initializer_list constructor and force customers to write their code "factory-style", as in

```
Roaring r = Roaring::bitmapOfList({3,4,,5})
```

which by the way is just as efficient.

File: tests/cpp_random_unit.cpp [hunk 5]
```diff
@@ -21,6 +21,7 @@
 //
 
 #include <assert.h>
+#include <inttypes.h>
```
Comment (kosak) [inline tests/cpp_random_unit.cpp:24]: Prefer native C++ header
```
#include <cinttypes>
```

Actually if you wanted you could

```
#include <cassert>
#include <cinttypes>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <ctime>
```

File: tests/cpp_random_unit.cpp (hunk 3 above)
Comment (lemire) [reply tests/cpp_random_unit.cpp:385]: Good catch. Fixing.

File: cpp/roaring.hh (hunk 2 above)
Comment (lemire) [reply cpp/roaring.hh:128]: Fixed in a commit.

File: tests/cpp_unit.cpp [hunk 6]
```diff
@@ -797,6 +814,25 @@ DEFINE_TEST(test_cpp_add_range_64) {
         assert_true(r1 == r2);
     }
 }
+DEFINE_TEST(test_bitmap_of_32) {
+        Roaring r1 =
+            Roaring::bitmapOf({1,2,4});
```
[the remark this answers is not in the document]
Comment (lemire) [reply tests/cpp_unit.cpp]: Fixed in a commit.

File: cpp/roaring.hh [hunk 7]
```diff
@@ -61,6 +61,15 @@ public:
         api::roaring_bitmap_add_many(&roaring, n, data);
     }
 
+    /**
```
[the remark this answers is not in the document]
Comment (lemire) [reply cpp/roaring.hh:64]: Done.

File: tests/cpp_random_unit.cpp (hunk 5 above)
Comment (lemire) [reply tests/cpp_random_unit.cpp]: Ok. Done.

File: tests/cpp_random_unit.cpp [hunk 8]
```diff
@@ -378,7 +379,7 @@ DEFINE_TEST(random_doublecheck_test_64) {
         const Roaring64Map &right = roars[rand() % NUM_ROARS];
 
 #ifdef ROARING_CPP_RANDOM_PRINT_STATUS
-        printf("[%lu]: %llu %llu %llu\n", step,
+        printf("[%lu]: %" PRIu64 " %" PRIu64 " %" PRIu64 "\n", step,
```
Comment (madscientist) [inline tests/cpp_random_unit.cpp:382]: In general of course this is portable.  However, MSVC didn't support C99 until the MSVC 2013 release, 9 years ago (and actually even that release didn't support all of it).  Maybe we don't care enough about supporting older MSVC and that's fine.

Comment (lemire) [reply tests/cpp_random_unit.cpp:382]: We require C11 in CRoaring.

File: cpp/roaring.hh (hunk 4 above)
Comment (lemire) [reply cpp/roaring.hh:73]: Ok. I will pull off constructors from this PR.

File: cpp/roaring.hh [hunk 9]
```diff
@@ -61,6 +61,15 @@ public:
...
+     * Construct a bitmap from an initializer list.
+     */
+    Roaring(std::initializer_list<uint32_t> l) : Roaring() {
+        for(uint32_t x : l) {
```
[the remark this answers is not in the document]
Comment (lemire) [reply cpp/roaring.hh:69]: This constructor was removed from this PR.

File: cpp/roaring.hh [hunk 10]
```diff
@@ -112,6 +121,18 @@ public:
         return ans;
     }
 
+    /**
+     * Construct a bitmap from a list of uint32_t values.
+     * E.g., bitmapOf({1,2,3}).
+     */
+    static Roaring bitmapOf(std::initializer_list<uint32_t> l) {
+        Roaring ans;
+        for(uint32_t x : l) {
```
[the remark this answers is not in the document]
Comment (lemire) [reply cpp/roaring.hh:131]: Good. Done.

File: cpp/roaring64map.hh [hunk 11]
```diff
@@ -51,6 +51,14 @@ public:
      */
     Roaring64Map(size_t n, const uint64_t *data) { addMany(n, data); }
 
+    /**
+     * Construct a bitmap from an initializer list.
+     */
+    Roaring64Map(std::initializer_list<uint64_t> l)  {
```
[the remark this answers is not in the document]
Comment (lemire) [reply cpp/roaring64map.hh:58]: Constructor removed from this PR.

**Comment by lemire** [comment]
I have addressed all issues that were raised, but I am leaving it up for further review (not merging right away).

File: cpp/roaring.hh [hunk 12]
```diff
@@ -112,6 +113,16 @@ public:
         return ans;
     }
 
+    /**
+     * Construct a bitmap from a list of uint32_t values.
+     * E.g., bitmapOf({1,2,3}).
```
Comment (kosak) [inline cpp/roaring.hh:118]: Typo (new spelling is bitmapOfList) and also spacing if you want

```
 * E.g., bitmapOfList({1, 2, 3}).
```

**Review by kosak — APPROVED** [review]
Looks great to me! There's one typo I found.

File: cpp/roaring.hh (hunk 12 above)
Comment (lemire) [reply cpp/roaring.hh]: Thanks. Fixed.

**Comment by Dr-Emann** [comment]
Wouldn't it be more idiomatic in c++ to make this a constructor, to allow construction like:

```c++
Roaring x{1, 2, 3, 4};
Roaring y = {1, 2, 3, 4};
auto z = Roaring64{1, 2, 3, 4};
```

File: cpp/roaring.hh (hunk 4 above)
Comment (Dr-Emann) [reply cpp/roaring.hh]: Do we ever foresee such a constructor with two integers?

I can't think of a case where I would think we'd ever want `Roaring x{5, 100}` to make anything _but_ a bitmap which contains only 5 and 100.

The common example of where such a constructor is problematic is the `std::vector<int> x{4, 2}` vs `std::vector<int>(4, 2)` (which create a vector containing 4 and 2, and 4 copies of 2 respectively). Honestly, I'd say if the vector api were being created now, the initializer list constructor would stay, and the fill constructor would go or be moved to a static method.

Comment (kosak) [reply cpp/roaring.hh]: Two args was just an example. The issue arises for constructors of any arity, so I think your best argument would have to assert that Roaring will never want a one-argument integer or pointer constructor either. But note, it actually *already has* conflicting one and two-arg constructors, to wit:

```
Roaring64Map x{0};   // this compiles today
```
The above currently invokes `Roaring64Map::Roaring64Map(roaring_bitmap_t *s)` with `nullptr`. WIth an initializer_list constructor the meaning would change, and the initializer_list constructor would be preferred.

And also
```
Roaring64Map a{0, 0};  // almost compiles today
```
and the only reason it *doesn't* compile is because there are *two* constructors to bind to and the compiler can't decide which one:

```
Roaring64Map::Roaring64Map(size_t n, const uint32_t *data)
Roaring64Map::Roaring64Map(size_t n, const uint64_t *data)
```

Yes these are nitpicky edge-case examples and calling those *particular* constructors with `nullptr` is not something anyone would want anyway. But this is different from saying they don't exist. My point is that it merits a discussion a little bigger than the scope of this PR.

TL;DR:
1. I assumed the scope of this particular PR was to provide an `initializer_list` alternative to the existing error-prone varargs `bitmapOf`. I think it has accomplished that and so this PR is a good quantum of change as is.
2. If my opinion was sought I'd say I had assumed `bitmapOf` was really only useful for unit tests, but if the community wanted the constructor I'd at least ask that such a change (along with an initializer_list `operator=` by the way) be put in its own PR so that the project owners can understand the pros and cons.

Comment (madscientist) [reply cpp/roaring.hh]: In modern C++ can't these be handled by creating ctors with std::nullptr_t types?  Differentiating between ctors that take a pointer versus an integer WRT NULL I think is a solved problem...?  I understand if the hope is to defer this discussion until later.

**Comment by lemire** [comment]
I will merge. We can add new constructors, etc. but I suggest we do it in a later release where we can discuss it fully.

File: cpp/roaring.hh (hunk 4 above)
Comment (kosak) [reply cpp/roaring.hh]: Ah, that's a good suggestion. Yes, I am just proposing we proceed with caution and with a discussion.

**Comment by lemire** [comment]
So I invite further pull requests in this direction.
